src/array.jl

```diff
@@ -19,6 +19,19 @@ function reftype(sz::Int)
     end
 end
 
+fixstringtype(T::Type) = T <: SubString || T === AbstractString ? String : T
+function fixstringtype(T::Union)
+    if isdefined(T, :a) && isdefined(T, :b)
+        Union{fixstringtype(T.a), fixstringtype(T.b)}
+    elseif isdefined(T, :a)
```

Comment:
why not just: `Union{map(fixstringtype, propertynames(T))...}`?

Reply:
And in general - is it possible to have only `:a` or only `:b` defined in `Union`? I thought they are both defined or both undefined.

Reply:
Actually I can't even find how to produce a `Union` with an undefined field now. How can I do that?

Reply:
`Union{}`

Reply:
But `Union{} isa Union` is `false`... :-D

Reply:
Oh - I did not know this

Reply:
I've simplified that function since we couldn't find a way to generate an incorrect `Union`.